x/concentrated-liquidity/pool_hooks_test.go

```diff
@@ -178,9 +178,136 @@ func (s *KeeperTestSuite) TestCallPoolActionListener() {
+//     e.g. if action prefix is "beforeSwap", transfer 1 token with denom "beforeSwap"
+//  2. Set this contract for all hooks defined by the test case (each case should have a list
+//     of action prefixes it wants to "activate")
+//  3. Run a series of actions that would trigger all the hooks (create, add to, withdraw from, swap against a position),
+//     and ensure that the correct denoms are in the account balance after each action/at the end.
+//
+// NOTE: we assume that set contracts have valid implementations for all hooks and that this is validated
+// at the contract setting stage at a higher level of abstraction. Thus, this class of errors is not covered
+// by these tests.
+func (s *KeeperTestSuite) TestPoolHooks() {
+	hookContractFilePath := "./testcontracts/compiled-wasm/hooks.wasm"
+
+	allBeforeHooks := []string{
+		before(types.CreatePositionPrefix),
+		before(types.WithdrawPositionPrefix),
+		before(types.SwapExactAmountInPrefix),
+		before(types.SwapExactAmountOutPrefix),
+	}
+
+	allAfterHooks := []string{
+		after(types.CreatePositionPrefix),
+		after(types.WithdrawPositionPrefix),
+		after(types.SwapExactAmountInPrefix),
+		after(types.SwapExactAmountOutPrefix),
+	}
+
+	testCases := map[string]struct {
+		actionPrefixes []string
+
+		expectedSetError error
+	}{
+		"single hook: before create position": {
+			actionPrefixes: []string{before(types.CreatePositionPrefix)},
+		},
+		"all before hooks": {
+			actionPrefixes: allBeforeHooks,
+		},
+		"all after hooks": {
+			actionPrefixes: allAfterHooks,
+		},
+		"all hooks": {
+			actionPrefixes: append(allBeforeHooks, allAfterHooks...),
+		},
+		"error: attempt to set a hook contract on an invalid action prefix": {
+			actionPrefixes: []string{"invalidActionPrefix"},
+			expectedSetError: types.InvalidActionPrefixError{
+				ActionPrefix: "invalidActionPrefix",
+				ValidActions: types.GetAllActionPrefixes(),
+			},
+		},
+	}
+
+	for name, tc := range testCases {
+		s.Run(name, func() {
+			s.SetupTest()
+			clPool := s.PrepareConcentratedPool()
+
+			// Upload and instantiate wasm code
+			rawCosmwasmAddress, cosmwasmAddressBech32 := s.uploadAndInstantiateContract(hookContractFilePath)
+
+			// Fund the contract with tokens for all action prefixes using a helper
+			for _, actionPrefix := range tc.actionPrefixes {
+				s.FundAcc(rawCosmwasmAddress, sdk.NewCoins(sdk.NewCoin(actionPrefix, sdk.NewInt(10))))
+			}
+
+			// Set the contract for all hooks as defined by tc.actionPrefixes
+			for _, actionPrefix := range tc.actionPrefixes {
+				// We use the bech32 address here since the set function expects it for security reasons
+				err := s.Clk.SetPoolHookContract(s.Ctx, validPoolId, actionPrefix, cosmwasmAddressBech32)
+
+				if tc.expectedSetError != nil {
+					s.Require().Error(err)
+					s.Require().ErrorContains(tc.expectedSetError, err.Error())
+					return
+				}
+
+				s.Require().NoError(err)
+			}
```

Comment:
It would be great to separate the test for `SetPoolHookContract` into a separate test fixtureto keep this test focused on hook triggers

Reply:
3eca42f0ad2222347c6700f5810bb734f3c39941